vortex-expr/src/analysis.rs

```diff
@@ -35,15 +35,20 @@ pub trait AnalysisExpr {
         None
     }
 
-    /// If an expression is returned, its value is an upper bound on the value of `expr`.
+    /// An expression for the upper non-null bound of this expression, if available.
```

Comment:
Why do we want this? returning null here implies no knowledge, it seems good that `max(lit(Scalar::null)) -> None` not `lit(Scalar::null)`.

I think this will be hard for people to impl correctly

Reply:
In develop there are three possible results of this method:
1. `None`
2. `Some(e)` where e evaluates to a non-null value.
3. `Some(e)` where e evaluates to a null value.

I think we very clearly defined the meaning of the first two (1: the max is unknown, 2: the max is this known value). The doc string does not clarify what the third case means. I am trying to clarify that in this docstring. I am not trying to forbid returning `None` when, at expression-creation-time, we know the max is unknown.

The specific use-case is that in Spiral, we have an array of file statistics. For some files, we know the maximum. For some files, we _do not_ know the maximum. In the latter case, the `max` column for that row is `null`.

Reply:
I see I think that when I read non-null I read non `None`. Maybe 
```suggestion
    /// An expression for the upper non-null scalar bound of this expression, if available.
```